ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java

```diff
@@ -169,202 +176,316 @@ public static KsqlEngine create(final KsqlConfig ksqlConfig) {
+  public List<QueryMetadata> execute(
+      final String sql,
       final KsqlConfig ksqlConfig,
-      final Map<String, Object> overriddenProperties) {
-    for (final String property : overriddenProperties.keySet()) {
-      if (IMMUTABLE_PROPERTIES.contains(property)) {
-        throw new IllegalArgumentException(
-            String.format("Cannot override property '%s'", property)
-        );
-      }
-    }
+      final Map<String, Object> overriddenProperties
+  ) {
+    final List<PreparedStatement<?>> statements = parseStatements(sql);
 
-    // Multiple queries submitted as the same time should success or fail as a whole,
-    // Thus we use tempMetaStore to store newly created tables, streams or topics.
-    // MetaStore tempMetaStore = new MetaStoreImpl(metaStore);
-    final MetaStore tempMetaStore = metaStore.clone();
+    final List<QueryMetadata> queries = doExecute(
+        statements, ksqlConfig, overriddenProperties, true);
 
-    // Build query AST from the query string
-    final List<PreparedStatement> queries = parseStatements(
-        queriesString,
-        tempMetaStore,
-        true
-    );
+    registerQueries(queries);
 
-    final List<QueryMetadata> queryMetadataList =
-        planQueries(queries, ksqlConfig, overriddenProperties, tempMetaStore);
-    return queryMetadataList;
+    return queries;
+  }
+
+  /**
+   * Determines if a statement is executable by the engine.
+   *
+   * @param statement the statement to test.
+   * @return {@code true} if the engine can execute the statement, {@code false} otherwise
+   */
+  public static boolean isExecutableStatement(final PreparedStatement<?> statement) {
+    return statement.getStatement() instanceof ExecutableDdlStatement
+        || statement.getStatement() instanceof QueryContainer
+        || statement.getStatement() instanceof Query;
   }
 
-  private List<QueryMetadata> planQueries(
-      final List<PreparedStatement> statementList,
+  private List<QueryMetadata> doExecute(
+      final List<? extends PreparedStatement<?>> statements,
       final KsqlConfig ksqlConfig,
       final Map<String, Object> overriddenProperties,
-      final MetaStore tempMetaStore
+      final boolean updateMetastore
   ) {
-    final List<PreparedStatement> preparedStatements = statementList.stream()
-        .map(statement -> statement.getStatement() instanceof QueryContainer
-            ? new PreparedStatement(
-            statement.getStatementText(),
-            ((QueryContainer) statement.getStatement()).getQuery())
-            : statement)
-        .filter(statement -> statement.getStatement() instanceof Query
-            || statement.getStatement() instanceof DdlStatement)
+    throwOnImmutableOverride(overriddenProperties);
+
+    // Multiple queries submitted as the same time should success or fail as a whole,
+    // Thus we use tempMetaStore to store newly created tables, streams or topics.
+    final MetaStore tempMetaStore = metaStore.clone();
```

Comment:
Do you need a temp metastore here? We're updating it for ddl statements from postProcessStatement, but what's the point of doing that?

Reply:
Existing functionality